### ISSUE TITLE
XML: add ability to provide configuration properties on the XMLInputFactory used by the Xml Parser

### ISSUE THREAD

**Issue by jphelp32** [akka/alpakka#1943]
### Short description

Add the ability to provide configuration properties on the XMLInputFactory used by the Xml Parser.

### Details

The biggest reason that this is necessary is b/c the default configuration of the aalto-xml parser does not disable external DTDs or DTD entity expansion, which is considered to be unsecure. At the very least, it should be possible to set the configuration properties needed to achieve secure XML parsing.

**Comment by ennru** [comment]
This would be good, indeed. We should even consider to disable external DTDs and DTD entity expansion by default in Alpakka.

**Comment by jphelp32** [comment]
Disabling those by default in alpakka would suffice for my interests.

I was a little unsure about how to approach this one, since the parser requires a specific factory (ie. aalto) rather than any generic XMLInputFactory. So a change similar to the one I made on the output side would not be appropriate.

**Comment by jphelp32** [comment]
@ennru would you like a separate issue and/or pull request just for the disabling of external DTDs and DTD entity expansion (short of full configurability of the input factory)?

**Comment by ennru** [comment]
An alternative could be to let the user pass an `XMLInputFactory => Unit` function. Or even `AsyncXMLInputFactory => Unit`.
And have some default implementation of it which disables the bad ones.

**Comment by jphelp32** [comment]
Something like the following:

```
@InternalApi private[xml] class StreamingXmlParser(ignoreInvalidChars: Boolean, configureXmlInputFactory: AsyncXMLInputFactory => Unit = _ => Unit) {
      ...
      private val factory: AsyncXMLInputFactory = new InputFactoryImpl()
      factory.setProperty(XMLInputFactory.SUPPORT_DTD, false)
      factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false)
      factory.setProperty(XMLConstants.FEATURE_SECURE_PROCESSING, true)   // I'm not sure about this one
      configureXmlInputFactory(factory)
      private val parser: AsyncXMLStreamReader[AsyncByteArrayFeeder] = factory.createAsyncFor(Array.empty)
      if (ignoreInvalidChars) {
        parser.getConfig.setIllegalCharHandler(new ReplacingIllegalCharHandler(0))
      }
      ...
}

**Comment by ennru** [comment]
I'd like to suggest to create a function in `XmlParsing` which does the settings we'd recommend as default.
```
object XmlParsing {
  val configureDefault: AsyncXMLInputFactory => Unit = { factory =>
    factory.setProperty(XMLInputFactory.SUPPORT_DTD, false)
    factory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false)
    factory.setProperty(XMLConstants.FEATURE_SECURE_PROCESSING, true)   // I'm not sure about this one
  }
```
That would be passed as default to a new overloaded variant of `parser`.

But it could be useful to even include the parser creation and make it `AsyncXMLInputFactory => AsyncXMLStreamReader[AsyncByteArrayFeeder]`!?

**Comment by ennru** [comment]
Fixed with #1982
